versioned/storage/cleanup/src/main/java/org/projectnessie/versioned/storage/cleanup/CutHistoryImpl.java

```diff
@@ -0,0 +1,150 @@
+import org.projectnessie.versioned.storage.common.objtypes.CommitObj;
+import org.projectnessie.versioned.storage.common.persist.CloseableIterator;
+import org.projectnessie.versioned.storage.common.persist.Obj;
+import org.projectnessie.versioned.storage.common.persist.ObjId;
+import org.projectnessie.versioned.storage.common.persist.Persist;
+
+public class CutHistoryImpl implements CutHistory {
+  private final CutHistoryParams context;
+  private final RateLimit scanRateLimiter;
+
+  public CutHistoryImpl(CutHistoryParams context, IntFunction<RateLimit> rateLimitIntFunction) {
+    this.context = context;
+    this.scanRateLimiter = rateLimitIntFunction.apply(context.scanObjRatePerSecond());
+  }
+
+  @Override
+  public CutHistoryScanResult identifyAffectedCommits() {
+    Builder result = CutHistoryScanResult.builder();
+
+    Persist persist = context.persist();
+    ObjId cutPoint = context.cutPoint();
+    result.cutPoint(cutPoint);
+
+    long numObjects = 0;
+    // Identify commits, whose "tails" overlap the cut point.
+    // scanAllCommitLogEntries() returns all commits in no specific order, parents may be scanned
+    // before or after their children.
+    try (CloseableIterator<Obj> scan = persist.scanAllObjects(Collections.singleton(COMMIT))) {
+      while (scan.hasNext()) {
+        scanRateLimiter.acquire();
+
+        CommitObj commit = (CommitObj) scan.next();
+        numObjects++;
+
+        var tail = commit.tail();
+        var cutPointIdx = tail.indexOf(cutPoint);
+        // Commits referencing the cut point as their last parent do not need to be rewritten
+        if (cutPointIdx >= 0 && cutPointIdx < tail.size() - 1) {
+          result.addAffectedCommitId(commit.id());
+        }
+      }
+
+      result.numScannedObjs(numObjects);
+      return result.build();
+    }
+  }
+
+  @Override
+  public CutHistoryResult cutHistory(CutHistoryScanResult scanResult) {
+    var result = CutHistoryResult.builder().input(scanResult);
+    var commitLogic = commitLogic(context.persist());
+    var cutPoint = context.cutPoint();
+
+    result.wasHistoryCut(false);
+    var failed = false;
+    for (ObjId id : scanResult.affectedCommitIds()) {
+      try {
+        var commitObj = commitLogic.fetchCommit(id);
+        checkState(commitObj != null, "Commit not found: %s", id);
+
+        if (commitObj.tail().size() <= 1) {
+          continue;
+        }
+
+        var tail = commitObj.tail();
+        var cutPointIdx = tail.indexOf(cutPoint);
+
+        if (cutPointIdx < 0 || cutPointIdx == tail.size() - 1) {
+          // may happen on re-invocation
+          continue;
+        }
+
+        var updatedCommitObj =
+            CommitObj.commitBuilder()
+                .from(commitObj)
+                .tail(commitObj.tail().subList(0, cutPointIdx + 1))
+                .build();
+
+        if (!context.dryRun()) {
+          context.persist().upsertObj(updatedCommitObj);
+          result.addRewrittenCommitId(id);
```

Comment:
Maybe always add `id` to the result, even in dry-run

Reply:
Why? These IDs can be obtained from `CutHistoryResult.input()`